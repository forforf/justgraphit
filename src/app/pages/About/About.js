
import React from 'react'
import Faq from 'Faq/Faq';
import './About.css';

export default React.createClass({
  render() {
    return (
      <div className="About">
        <h3>JustGraphIt</h3>
        <h4>FAQ</h4>
        <Faq />

        <ul>
          <li>
            version: 0.3.0
          </li>
          <li>
            author email: <a href="mailto: dmarti21@gmail.com">dmarti21@gmail.com</a>
          </li>
          <li>
            license: MIT
          </li>

        </ul>
      </div>
    )
  }
})
