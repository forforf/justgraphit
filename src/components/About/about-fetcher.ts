/*
Although this is a bit more complex than it needs to be, it's more an exercise
about using functional code to separate concerns.

What the code does:
Given a list of constant names (like ["LICENSE", "VERSION"])
look up the corresponding CRA env variable in process.env.
Return an object with the env variables values e.g.:
  {LICENSE: "MIT", VERSION: "0.0.1"}

So why didn't I use something like:
```
function getProjectData(inputNames) {

  let envVarObj = {}
  inputNames.forEach(name => {
    const envVarName = `REACT_APP_${name}`;
    envVarObj[name] = process.env[envVarName];
  }
  return envVarObj;
}
```

* process.env is a "magical" operator. It just suddenly appears.
* process.env is tightly coupled with getProjectData even though they have separate concerns.
* mocking process.env for testing requires jumping through some hoops
* the forEach loop is mutating envVarObj as a side effect

What I wanted:
* separate the concerns of "process.env" (the provider) from the business logic of getProjectData which
  transforms a list of names into an object with the names as the key of that object, and the values
  are calculated based on the names (note that the business logic here doesn't care whether process.env does
  the calculation or not).
* no mutations (i.e. imuttable)
* dependency injection (i.e. no magical appearances)

How I did it:
Given:
* a concrete "provider" (a thin wrapper around process.env in this case,
  that also takes care of the inputName -> envVarName transform.
* a "hydrator" reducer that returns a reducing function that immutably creates an object based on
  input keys, using an *abstract* provider to calculate the values based on the provided keys.
  Note: reducer in the javascript array.reduce function, not as in map/reduce.
* A a "key list" of input names that will be the keys of the returned object.

Do: `keyList.reduce(hydrator(provider), {})`
* the provider is injected into the hydrator reducing function
* us reduce to convert the key list array into an object using the hydrator function

*/

function getEnvVar(envVar: string): string {
  const reactEnvVar = `REACT_APP_${envVar}`;
  return process.env[reactEnvVar] ?? '';
}

// hydrator, given a provider, returns the reducing function that
// that uses that provider to build key-value pairs
function hydrator(provider: (s: string) => string) {
  // returns the reducing function with the 'provider' dependency injected in the closure
  return function reducer(
    current_obj: Record<string, string>,
    current_key: string
  ) {
    // combine the current object with the key-value of current_key: provider(current_key)
    return { ...current_obj, [current_key]: provider(current_key) };
  };
}

// By passing a provider as an argument, we can swap out how the project data is configured easily
// plus it makes for easy testing
export function getProjectData(
  keyList: string[],
  provider = getEnvVar
): Record<string, string> {
  // take the keys and uses provider to hydrate the object
  return keyList.reduce(hydrator(provider), {});
}

// Note: hydrator is not being injected (so it is kind of "magically appearing).
// This could be fixed by adding a hydrator argument to getProjectData, but ... I didn't feel like going that far.
