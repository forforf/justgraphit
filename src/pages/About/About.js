
import React from 'react'
import Faq from '../../components/Faq/Faq';

export default React.createClass({
  render() {
    return (
      <div>
        <h3>JustGraphIt</h3>
        <h4>FAQ</h4>
        <Faq />

        <ul>
          <li>
            version: 0.3.0
          </li>
          <li>
            author email: <a href="mailto: dmarti21@gmail.com">forforf c/o Dave Martin</a>
          </li>
          <li>
            license: MIT
          </li>

        </ul>
      </div>
    )
  }
})
