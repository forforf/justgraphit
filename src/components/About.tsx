import React from 'react';
import Faq from './Faq';
import './About/About.css';

const About = (): JSX.Element => (
  <div className="About">
    <h3>App Name: {process.env.REACT_APP_NAME}</h3>
    <h4>FAQ</h4>
    <Faq />
    <ul>
      <li>version: {process.env.REACT_APP_VERSION}</li>
      <li>
        author email:
        <a href="mailto: dmarti21@gmail.com">dmarti21@gmail.com</a>
      </li>
      <li>license: MIT</li>
    </ul>
  </div>
);

export default About;
