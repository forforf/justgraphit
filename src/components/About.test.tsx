import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow'
import About from './About';

const OLD_ENV = process.env;

beforeEach(() => {
  jest.resetModules() // Most important - it clears the cache
  process.env = { ...OLD_ENV }; // Make a copy
});

afterAll(() => {
  process.env = OLD_ENV; // Restore old environment
});

test('deep snapshot', () => {
  const component = renderer.create(
    <About />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('shallow snapshot', () => {
  const renderer = ShallowRenderer.createRenderer();
  renderer.render(<About />);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot()
});
