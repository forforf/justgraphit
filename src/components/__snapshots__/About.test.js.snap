// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep snapshot 1`] = `
<div
  className="About"
>
  <h3>
    App Name: 
    IDE-mock-name
  </h3>
  <h4>
    FAQ
  </h4>
  <ul
    className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
  >
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Where is my data stored?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          The data is stored on the local device, the network never sees it.
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          Do I need a network connection?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          Only to get to the initial page, after that no network is required.
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          How much is it?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          Free.
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          I don't see any ads?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          No Ads.
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          How was this built?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          Originally on Angular1, then Angular2, then classic React, now functional React
        </p>
      </div>
    </li>
    <li
      className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
      disabled={false}
    >
      <div
        className="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
      >
        <span
          className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
        >
          How do I use it?
        </span>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
        >
          Enter a number you want plotted on the graph page. That's it. To create a new graph, just change the name.
        </p>
      </div>
    </li>
  </ul>
  <ul>
    <li>
      version: 
      IDE-mock-versions
    </li>
    <li>
      author email: 
      <a
        href="mailto: dmarti21@gmail.com"
      >
        dmarti21@gmail.com
      </a>
    </li>
    <li>
      license: MIT
    </li>
  </ul>
</div>
`;

exports[`shallow snapshot 1`] = `
<div
  className="About"
>
  <h3>
    App Name: 
    IDE-mock-name
  </h3>
  <h4>
    FAQ
  </h4>
  <Faq />
  <ul>
    <li>
      version: 
      IDE-mock-versions
    </li>
    <li>
      author email: 
      <a
        href="mailto: dmarti21@gmail.com"
      >
        dmarti21@gmail.com
      </a>
    </li>
    <li>
      license: MIT
    </li>
  </ul>
</div>
`;
